lib/platform/linux-uart.c

```diff
@@ -0,0 +1,432 @@
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a
+ * copy of this software and associated documentation files (the "Software"),
+ * to deal in the Software without restriction, including without limitation
+ * the rights to use, copy, modify, merge, publish, distribute, sublicense,
+ * and/or sell copies of the Software, and to permit persons to whom the
+ * Software is furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included
+ * in all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS
+ * OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+ * THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+ * OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+ * ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+ * OTHER DEALINGS IN THE SOFTWARE.
+ *
+ */
+
+#ifdef __linux__
+
+#include "../switchtec_priv.h"
+#include "switchtec/switchtec.h"
+#include "gasops.h"
+
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <sys/ioctl.h>
+#include <sys/mman.h>
+#include <fcntl.h>
+#include <unistd.h>
+#include <errno.h>
+#include <signal.h>
+#include <stddef.h>
+#include <assert.h>
+#include <string.h>
+
+#include <sys/file.h>
+#include <termios.h>
+#include "../crc8.h"
+
+static const char *dbginfo_off = "pscdbg 0 all\r";
+static const char *echo_off = "echo 0\r";
+static const char *echo_on = "echo 1\r";
+
+struct switchtec_uart{
+	struct switchtec_dev dev;
+	int fd;
+};
+
+#define to_switchtec_uart(d) \
+	((struct switchtec_uart *) \
+	 ((char *)d - offsetof(struct switchtec_uart, dev)))
+
+#define UART_MAX_WRITE_BYTES			(100)
+#define RETRY_NUM				(10000)
+
+enum switchtec_uart_cmd_type {
+	SWITCHTEC_UART_CONTROL = 0,
+	SWITCHTEC_UART_GAS_READ = 1,
+	SWITCHTEC_UART_GAS_WRITE = 2,
+};
+
+static int send_cmd(int fd, const char *str, enum switchtec_uart_cmd_type type,
+		    uint32_t addr, size_t n, const void *src, uint8_t crc)
+{
+	int ret;
+	int i;
+	char cmd[1024];
+	char *b = cmd;
+
+	switch (type) {
+		case SWITCHTEC_UART_CONTROL:
+			memcpy(cmd, str, strlen(str));
+		break;
+		case SWITCHTEC_UART_GAS_WRITE:
+			b += snprintf(cmd, sizeof(cmd),
+				      "gaswr -c -s 0x%x 0x", addr);
+			for (i=0; i<n; i++)
+				b += snprintf(b, sizeof(cmd),
+					      "%02x", *(uint8_t *)(src + n - 1 - i));
+			b += snprintf(b, sizeof(cmd), " 0x%x\r", crc);
+		break;
+		case SWITCHTEC_UART_GAS_READ:
+			b += snprintf(cmd, sizeof(cmd),
+				      "gasrd -c -s 0x%x %zu\r", addr, n);
+		break;
+	}
+
+	ret = write(fd, cmd, strlen(cmd));
+	if (ret != strlen(cmd))
+		return -1;
+	else
+		return 0;
+}
+
+static int read_resp_line(int fd, char *str)
+{
+	int ret;
+	int i;
+	int cnt = 0;
+
+	for (i = 0; i < RETRY_NUM; i++) {
```

Comment:
So RETRY_NUM is 10000 and the timeout on read() is 0.5s. So if the device doesn't respond it takes 5000s (more than an hour) before this function returns with a failure.

I think you should have a while(1) and then break if read() ever returns 0. Then increase the timeout to maybe 5s. That way if we don't receive anything for the timeout period we return failure. Otherwise if the response is large and divided into 1000 chunks (unlikely) the function will still be correct.

Reply:
update some data: (history for RETRY_NUM)
1. first version with 5000us delay, the RETRY_NUM is 100
it is generated by following time consuming mrpc command operations:
a1.firmware toggle take about 5 loops
b1.status/gui/bw, about 17 loops
c1.log-dump/fw read, about 29 loops

2. after remove the 5000us delay, the RETRY_NUM change to 10000, because the loop number collect when test also increased:
a2.firmware toggle take about 150 loops
b2.status/gui/bw, about 1800 loops
c2.log-dump/fw read, about 3200  loops

not dig into why the proportion between b1/c1 and b2/c2 keep almost no change
but a1/b1 and a2/b2 changes a lot. Maybe the first time log number into excel with some error. 

Anyway.  Yes, more than 1 hour is not acceptable.
change to code to 5ms timeout, cause fw-toggle is very time consuming, when read return 0, it may still fw in doing toggle action, event not in respone stage.

Reply:
to 5s, typo.